### PR TITLE
Allowing Netty TLS bootstrap handler to be sharable.

### LINKED ISSUE #504: Netty throws exception when creating > 1 TLS transport.
When attempting to create multiple TLS transports from the same builder, Netty throws an exception:

```
io.netty.channel.ChannelPipelineException: io.grpc.transport.netty.ProtocolNegotiators$SslBootstrapHandler is not a @Sharable handler, so can't be added or removed multiple times.
	at io.netty.channel.DefaultChannelPipeline.checkMultiplicity(DefaultChannelPipeline.java:564)
	at io.netty.channel.DefaultChannelPipeline.addFirst0(DefaultChannelPipeline.java:114)
	at io.netty.channel.DefaultChannelPipeline.addFirst(DefaultChannelPipeline.java:108)
	at io.netty.channel.DefaultChannelPipeline.addFirst(DefaultChannelPipeline.java:291)
	at io.netty.channel.DefaultChannelPipeline.addFirst(DefaultChannelPipeline.java:246)
	at io.grpc.transport.netty.ProtocolNegotiators$AbstractBufferingHandler.channelRegistered(ProtocolNegotiators.java:192)
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelRegisteredNow(ChannelHandlerInvokerUtil.java:32)
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRegistered(DefaultChannelHandlerInvoker.java:50)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRegistered(AbstractChannelHandlerContext.java:114)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRegistered(DefaultChannelPipeline.java:833)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:487)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$100(AbstractChannel.java:401)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:461)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:322)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:356)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:703)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
	at java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by nmittler**
Also adding better handling for failures that occur during the protocol negotiation. Now properly catching the error, closing the channel, and notifying the application that the transport has terminated.

Fixes #504

### CODE REVIEW

**Comment by nmittler** [comment]
@ejona86 PTAL

File: testing/src/main/java/io/grpc/testing/TestUtils.java [hunk 1]
```diff
@@ -62,6 +65,7 @@
  * Common utility functions useful for writing tests.
  */
 public class TestUtils {
+  public static final String TESTCA_HOST = "foo.test.google.fr";
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/TestUtils.java]: This isn't the test CA host, but more of test server host. Before it didn't matter too much, but this change would make it part of our API, so we would want it to be named accurately.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java [hunk 2]
```diff
@@ -86,49 +87,59 @@ public static ProtocolNegotiator tls(final SslContext sslContext,
     Preconditions.checkNotNull(sslContext, "sslContext");
     Preconditions.checkNotNull(inetAddress, "inetAddress");
 
-    final ChannelHandler sslBootstrapHandler = new ChannelHandlerAdapter() {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: FYI, you could have left the class defined within the method and just given it a name (so no longer anonymous). I know it is rare to see non-anonymous classes defined in methods though.

**Comment by ejona86** [comment]
I would have preferred the error handling in a separate PR, as fixing the tls bootstrap would have been a low-risk fix but the error handling is much more invasive. It's going to take me a bit to look over the error handling.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java (hunk 2 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: Ah right ... not sure I've ever seen it done in real code but I do recall it being possible.

File: testing/src/main/java/io/grpc/testing/TestUtils.java (hunk 1 above)
Comment (nmittler) [inline testing/src/main/java/io/grpc/testing/TestUtils.java]: Agreed ... done.

**Comment by nmittler** [comment]
@ejona86 I'll separate out the error handling ... stand by :)

**Comment by nmittler** [comment]
@ejona86 IIRC the reason I also made the error handling changes was because this error was not being detected, so I really have no way to test the change.  When errors happen during negotiation, the transport terminated handler doesn't seem to be called.

File: netty/build.gradle [hunk 3]
```diff
@@ -5,7 +5,18 @@ dependencies {
...
     // Tests depend on base class defined by core module.
-    testCompile project(':grpc-core').sourceSets.test.output
+    testCompile project(':grpc-core').sourceSets.test.output,
+                project(':grpc-testing')
+}
+
+test {
+    jvmArgs "-Xbootclasspath/p:" + configurations.alpnboot.asPath
+    testLogging {
+        exceptionFormat "full"
```
Comment (ejona86) [inline netty/build.gradle]: I know that exceptionFormat is already being set to "full" in the root build.gradle. It seems like if these other options are useful, we would want them at the root as well.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java [hunk 4]
```diff
@@ -159,23 +175,27 @@ public void start(Listener transportListener) {
...
   }
 
+  /**
+   * Called after the channel has been closed to notify the application that the transport
+   * has been terminated.
+   */
+  private void notifyClosed(ChannelFuture closeFuture) {
+    if (!closeFuture.isSuccess()) {
+      // The close failed, just use the result of the future.
+      failureCause = closeFuture.cause();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java]: This can overwrite the failureCause; it seems we would want to check it for `null` since the handshake could have failed.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java [hunk 5]
```diff
@@ -112,7 +116,19 @@
     }
 
     handler = newHandler();
-    negotiationHandler = negotiator.newHandler(handler);
+    FailureListener closeOnNegotiationFailure = new FailureListener() {
+      @Override
+      public void negotiationFailed(ChannelHandlerContext ctx, Throwable cause) {
+        failureCause = cause;
+        ctx.close().addListener(new ChannelFutureListener() {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java]: It looks like we need the listener on closeFuture independent of whether negotiation failed. Calling ctx.close() here can be appropriate, but the listener itself should probably be added where it used to.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java [hunk 6]
```diff
@@ -86,49 +87,59 @@ public static ProtocolNegotiator tls(final SslContext sslContext,
...
-        ctx.pipeline().replace(this, "sslHandler", sslHandler);
-      }
-    };
+    final SslBootstrapHandler sslBootstrap = new SslBootstrapHandler(sslContext, inetAddress);
     return new ProtocolNegotiator() {
       @Override
-      public Handler newHandler(Http2ConnectionHandler handler) {
-        return new BufferUntilTlsNegotiatedHandler(sslBootstrapHandler, handler);
+      public Handler newHandler(FailureListener failureListener, Http2ConnectionHandler handler) {
+        return new BufferUntilTlsNegotiatedHandler(failureListener, sslBootstrap, handler);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: Would it be better to create the SslHandler within newHandler()? It seems we don't really gain anything by using SslBootstrapHandler.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java [hunk 7]
```diff
@@ -236,15 +260,15 @@ protected void failBufferedAndClose(ChannelHandlerContext ctx) {
...
       }
-      /**
-       * In case something goes wrong ensure that the channel gets closed as the
-       * NettyClientTransport relies on the channel's close future to get completed.
-       */
-      ctx.close();
+
+      // Notify the application that the protocol negotiation has failed.
+      if (!previouslyFailed) {
+        listener.negotiationFailed(ctx, cause);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: I'd feel a lot better if ctx.close() was also called here.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java [hunk 8]
```diff
@@ -224,10 +245,13 @@ public void flush(ChannelHandlerContext ctx) {
 
     @Override
     public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Exception {
-      failBufferedAndClose(ctx);
+      fail(ctx, new Exception("Channel closed while performing protocol negotiation"));
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: The old code would call the approximate equivalent of super.close here. Closing within this class is important, as it is very hard to track the closure back to the negotiationFailed implementation.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+  @Test
+  public void creatingMultipleTlsTransportsShouldSucceed() throws Exception {
+    int port = TestUtils.pickUnusedPort();
+    InetSocketAddress address = TestUtils.overrideHostFromTestCa("localhost", port);
+
+    File serverCert = TestUtils.loadCert("server1.pem");
+    File key = TestUtils.loadCert("server1.key");
+    SslContext serverContext = GrpcSslContexts.forServer(serverCert, key).build();
+    NettyServer server = new NettyServer(address, NioServerSocketChannel.class,
+            new NioEventLoopGroup(),new NioEventLoopGroup(), serverContext, 100,
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Add a space after the comma.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+          @SuppressWarnings("unchecked")
+          public void transportTerminated() {
+            futures[ix].set(null);
+          }
+        });
+      }
+
+      // Give the transports a moment to complete negotiation before we tear them down. If we
+      // try closing the transports during negotiation, the failureCause will be non-null.
+      Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Does it really need to be uninterruptable?

File: netty/build.gradle (hunk 3 above)
Comment (nmittler) [inline netty/build.gradle]: done.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 10 above)
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Doing an RPC on each transport seems better than sleeping.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 9 above)
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: You have to shut down the event loop groups, otherwise they continue running.

Can you specify the argument `1` to the loops so tons of threads aren't created? Also, feel free to reuse the event loop between the boss and workers.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java (hunk 5 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java]: Acknowledged ... solution discussed offline.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java (hunk 4 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java]: done.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java (hunk 8 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: done ... fail now calls `ctx.close()`

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java (hunk 7 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: done.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 10 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Reorganized the test ... there is no longer a sleep.

**Comment by nmittler** [comment]
@ejona86 PTAL

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 9 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: done.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java [hunk 11]
```diff
@@ -86,36 +90,46 @@ public static ProtocolNegotiator tls(final SslContext sslContext,
...
+
+    SslBootstrapHandler(SslContext sslContext, InetSocketAddress address) {
+      this.sslContext = sslContext;
+      this.address = address;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+      // TODO(nmittler): This method is currently unsupported for OpenSSL in Netty < 4.1.Beta6.
+      SSLEngine sslEngine = sslContext.newEngine(ctx.alloc(),
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: Would it be better to create the SslHandler within newHandler()? It seems all this code could be within newHandler() and we pass an sslHandler to BufferUntilTlsNegotiatedHandler instead of SslBootstrapHandler.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java [hunk 12]
```diff
@@ -158,6 +172,7 @@ public Handler newHandler(Http2ConnectionHandler handler) {
     private Queue<ChannelWrite> bufferedWrites = new ArrayDeque<ChannelWrite>();
     private boolean writing;
     private boolean flushRequested;
+    private Throwable failureCause;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: It seems like we don't really gain much by saving this Throwable.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java [hunk 13]
```diff
@@ -224,18 +244,24 @@ public void flush(ChannelHandlerContext ctx) {
...
-      if (bufferedWrites != null) {
-        Exception e = new Exception("Buffered write failed.");
-        while (!bufferedWrites.isEmpty()) {
-          ChannelWrite write = bufferedWrites.poll();
-          write.promise.setFailure(e);
+    protected final void fail(ChannelHandlerContext ctx, Throwable cause) {
+      if (failureCause == null) {
+        failureCause = cause;
+        if (bufferedWrites != null) {
+          Exception e = new Exception("Buffered write failed.");
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: Use `cause` instead?

That would also allow passing in a Status Exception which would allow us to set a Status other than UNKNOWN.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Tsests for {@link NettyClientTransport}.
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: s/Tsests/Tests/

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+
+  private InetSocketAddress address;
+  private NettyServer server;
+  private List<NettyClientTransport> transports;
+  private List<NioEventLoopGroup> groups;
+
+  @Before
+  public void setup() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    groups = new ArrayList<NioEventLoopGroup>();
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Initialize these at the declaration?

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+      stream.halfClose();
+    }
+
+    // Wait for the RPCs to complete.
+    for (SettableFuture rpcFuture : rpcFutures) {
+      rpcFuture.get(10, TimeUnit.SECONDS);
+    }
+  }
+
+  private NioEventLoopGroup newGroup() {
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Why do you need multiple groups? Can't you just have a single one and share it between all usages?

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java (hunk 11 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: We can move the creation of the bootstrap handler to the `newHandler` method, but we still need it because we need the `ChannelHandlerContext.alloc()`.

Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: Oh, right... _That's_ why we had it to begin with. Okay, it makes sense as-is, then.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java (hunk 12 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: done.

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java (hunk 13 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: done.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 14 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Not liking my creative spelling? :)

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 15 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: done.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 16 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: done.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 14 above)
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: It did have a very nice symmetry. Also the difference in pronunciation wouldn't be too noticeable. We should try to get Webster to add it to the dictionary and then change all instances of the outdated and quaint "test" spelling. :)

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+        }
+      };
+    }
+
+    @Override
+    public void serverShutdown() {
+    }
+  }
+
+  private static class StringMarshaller implements Marshaller<String> {
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Use Metadata.ASCII_STRING_MARSHALLER instead?

**Comment by nmittler** [comment]
@ejona86 PTAL

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java [hunk 18]
```diff
@@ -181,8 +183,13 @@ public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
...
 
     @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+      fail(ctx, cause);
+    }
+
+    @Override
     public void channelInactive(ChannelHandlerContext ctx) throws Exception {
-      failBufferedAndClose(ctx);
+      fail(ctx, new Exception("Connection broken while performing protocol negotiation"));
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: Use Status.asRuntimeException() for this and similar Exceptions?

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 17 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Not sure I follow ... that seems to implement the `AsciiMarshaller` interface, not `Marshaller`.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+  private final List<NettyClientTransport> transports = new ArrayList<NettyClientTransport>();
+  private NioEventLoopGroup group;
+  private InetSocketAddress address;
+  private NettyServer server;
+
+  @Before
+  public void setup() throws Exception {
+    MockitoAnnotations.initMocks(this);
+
+    transports.clear();
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Why clear? We know nothing is in it, and even if there is we will want to clean it up during teardown().

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 17 above)
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Oh, nm. Sorry. This isn't a Metadata.*Marshaller. Duh.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 19 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: done.

**Comment by ejona86** [comment]
@nmittler LGTM! Thanks!

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java (hunk 18 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: So the question is which status to use in each case:

1. channelInactive (unavailable?)
2. close (cancelled?)
3. http2 plaintext upgrade failed (failed_precondition?)

Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: channelInactive. UNAVAILABLE seems reasonable.

close. No, not cancelled, since even if we willingly closed the connection that would just make it UNAVAILABLE. INTERNAL or UNAVAILABLE sounds fine to me. If in real doubt we can do UNKNOWN.

http2 plaintext upgrade failed. No, not FAILED_PRECONDITION; UNAVAILABLE is strictly better here (see the litmus test in FAILED_PRECONDITION's doc). UNAVAILABLE does seem most appropriate, since I think that is the same status we should give (but we may not be doing so) if hostname lookup fails or can't create a TCP connection.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+            group, group, serverContext, 100, DEFAULT_WINDOW_SIZE, DEFAULT_WINDOW_SIZE);
+    server.start(new TestServerListener());
+  }
+
+  @After
+  public void teardown() throws Exception {
+    for (NettyClientTransport transport : transports) {
+      transport.shutdown();
+    }
+    transports.clear();
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Why is this needed?

File: netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java (hunk 18 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java]: done.

File: netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java (hunk 20 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java]: Right, JUnit creates a new instance of the classbefore calling each test method ... it just seems weird to not treat the setup method as a the place to make sure the test is in a clean state.  Anyway .. done.

**Comment by nmittler** [comment]
@ejona86 PTAL ... shall I cherrypick?

**Comment by ejona86** [comment]
@nmittler still LGTM.

Yeah, let's cherrypick. Not for tls handler stuff as much for the negotiation error stuff.

**Comment by nmittler** [comment]
cherry-picked as 4ee2a65
